src/main/scala/org/uaparser/scala/OS.scala

```diff
@@ -32,16 +32,20 @@ object OS {
         osReplacement
           .map(replacementBack1(matcher))
           .orElse(matcher.groupAt(1)).map { family =>
-            val major = v1Replacement
-              .map(replacementBack1(matcher))
-              .orElse(matcher.groupAt(2))
-            val minor = v2Replacement.orElse(matcher.groupAt(3))
-            val patch = v3Replacement.orElse(matcher.groupAt(4))
-            val patchMinor = v4Replacement.orElse(matcher.groupAt(5))
+            val major = v1Replacement.flatMap(replace(matcher)).orElse(matcher.groupAt(2))
+            val minor = v2Replacement.flatMap(replace(matcher)).orElse(matcher.groupAt(3))
+            val patch = v3Replacement.flatMap(replace(matcher)).orElse(matcher.groupAt(4))
+            val patchMinor = v4Replacement.flatMap(replace(matcher)).orElse(matcher.groupAt(5))
             OS(family, major, minor, patch, patchMinor)
         }
       }
     }
+
+    private def replace(matcher: Matcher)(replacement: String): Option[String] =
```

Comment:
Do you mind moving this up to the level of `replacementBack1`, which has similar functionality? Also might as well make it `private[this]` for consistency.

Reply:
Also, wouldn't it be better for this method to ignore e.g. `$a` rather than throw an exception?

Reply:
@travisbrown Thank you so much
I have updated. PTAL